src/solver/impls/imex-bdf2/imex-bdf2.cxx

```diff
@@ -651,9 +652,15 @@ void IMEXBDF2::constructSNES(SNES* snesIn) {
       // Create data structure for SNESComputeJacobianDefaultColor
       MatFDColoringCreate(Jmf, iscoloring, &fdcoloring);
       // Set the function to difference
-      MatFDColoringSetFunction(
-          fdcoloring, reinterpret_cast<PetscErrorCode (*)()>(FormFunctionForColoring),
-          this);
+#if PETSC_VERSION_GE(3, 24, 0) \
+    || (PETSC_VERSION_GE(3, 23, 0) && PETSC_VERSION_RELEASE == 0)
+#define FUNC_MAYBE_CAST(func) func
+#else
+#define FUNC_MAYBE_CAST(func) reinterpret_cast<MatFDColoringFn>(func)
+#endif
+      MatFDColoringSetFunction(fdcoloring, FUNC_MAYBE_CAST(FormFunctionForColoring),
+                               this);
+#undef FUNC_MAYBE_CAST
```

Comment:
Do we need all this? Could we at least put it in the petsc interface header?

Reply:
We could always call the cast, but that might hide issues, so I would like to avoid that.

I put it here, as it is only used here, but we can of course also move it to a header ...